Update Creating a Module guide (#6048)
aces/Loris: docs/wiki/99_Developers/Creating_Module.md

# Creating a Module

## Overview

A Loris module consists of a group of pages that perform a specific function or assist with a specific task. Examples of modules include `Document Repository`, `Candidate List (Access Profiles)`, `Imaging Browser` ...

## Directory structure

Create a new folder corresponding to your module name under `modules/`. 
This name should be short and descriptive. 
The name should use only lowercase alphanumeric characters and underscores, e.g. `document_repository`.
The directory name correponds to the URL that LORIS will use to load your module.
For the remainder of this document we will assume your new module is named `my_new_module`.

### PHP Files

The following files are required for the module to load in LORIS:

* A PHP file called "module": `php/module.class.inc`.
* A PHP file with the same name as the module: `php/my_new_module.class.inc`.

The namespace must be set to `LORIS\$modulename` for anything defined in the PHP directory. Filenames need to be lowercase in order to be autoloaded. Class names can be any case.

> The remaining PHP files will depend heavily on the kind of module you are creating. 

Reviewer: ```suggestion
The remaining PHP files will depend heavily on the kind of module you are creating. See the Imaging Browser or DICOM Archive for examples of modules with subpages. 
```

The most common type of LORIS module will extend the class `NDB_Page` and belong to one of two types:
"Menu Filters" and "Forms".

#### Menu Filters

This is a type of page that displays a table of data with options to filter the values of this table.

The correct way to create a menu filter is to include additional classes that represent the idea of a
row in the table displayed to a user. This differs from a row in the database itself; for example, a row in the 
menu filter module may represent a joining of columns across many database tables.

If you are creating a Menu Filter page, your class `php/my_new_module.class.inc` should extend
the class `\DataFrameworkMenu`.

[See this file for an example](../../../modules/dicom_archive/php/dicom_archive.class.inc).

##### Data Instance
You should create a class `php/mynewmodulerow.class.inc` to represent a Menu Filter row. This will

> provide basic structure for the row and include permission flags that can restrict data to users

Reviewer 2: I don't know what "permission flags" are

> provide basic structure for the row and include permission flags that can restrict data to users

Author: ```suggestion
provide basic structure for the row and will likely restrict data access to users
```

affiliated with a given Site or Project.

[See this file for an example](../../../modules/dicom_archive/php/dicomarchiverow.class.inc).

##### Row Provisioner.
This file should be accompanied by a class `php/mynewmodulerowprovisioner.class.inc`. Provisioners usually
contain the SQL statement(s) used to query the database for data that will be consolidated into a

> Menu Filter row.

Reviewer 2: This should clarify that data can be provisioned by any source, and doesn't need to come from an SQL database.

Author: I don't really understand what this means -- doesn't all data in LORIS come from MySQL? Could you maybe use the suggest feature to make the changes you'd like to see?

of updating the relevant tables in LORIS.

#### Creating a patch

The files referenced above will be run automatically for new LORIS installations. In addition to these files,
you must create a "patch" file that existing LORIS instances can run in order to use your new module.

Create a file `SQL/New_patches/$date_AddMyNewModule.sql`. This file should contain the same commands you added
to `SQL/0000-00-01-Permission.sql` and `SQL/0000-00-02-Modules.sql`.

### Accessing the module

The new module must be included in the modules table 
Once you created the modules as described above, go to `$LORIS$/$ModuleName$` to see your newly created page

### Testing

Your module must contain a file `test/TestPlan.md` that enumerates a sequence of actions that a developer can
follow to test all functionalities of your module.

#### Automated Testing

We strongly encourage that new modules include integration tests. These must be written in the file
`test/my_new_moduleTest.php`.

### Documentation

A new module must contain a `README.md` file that briefly describes the functionality of the module.
Examples can be found in any existing module.

> It must also contain help text that will be displayed to users. In contrast to the README, this should

Reviewer 2: This should say where the help text document(s) go.

> It must also contain help text that will be displayed to users. In contrast to the README, this should

Author: ```suggestion
The new module must also a markdown file under `help/` that will be displayed to users. In contrast to the README, this should
```
